### PR TITLE
Consolidate file downloading utilities

### PR DESCRIPTION
**Pull request by miurahr**
Consolidate http connection utility methods into org.omegat.util.net.HttpConnectionUtils class.

There are several similar functions in WikiGet, StaticUtils and DictionaryManager.

This is also a preparation for plugin file download feature for #69.

Signed-off-by: Hiroshi Miura <miurahr@linux.com>

### CODE REVIEW

**Comment by amake** [comment]
Thanks, I think this is a great improvement.

However I'm concerned about one thing: I think third-party plugins may use some of the old public methods, so rather than removing them I think we should deprecated them and keep them for at least one release.

**Comment by miurahr** [comment]
> However I'm concerned about one thing: I think third-party plugins may use some of the old public methods,

I'm also afraid that. There seems to be no clean definitions to open API which one is usable and stable for plugins.
IMHO, now is a good chance to define a clean API for plugins and encourage plugins developer update their products.

Can we define a java package such as `org.omegat.openapi` that is usable for plugins? 

> so rather than removing them I think we should deprecate them and keep them for at least one release.

Ok, I'd like to do it.

File: src/org/omegat/util/net/HttpConnectionUtils.java [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+     */
+    private static final int TIMEOUT_MS = 10_000;
+
+    /**
+     * Don't instanciate util class.
+     */
+    private HttpConnectionUtils() {}
+
+    /**
+     * Download a file to memory
```
Comment (miurahr) [inline src/org/omegat/util/net/HttpConnectionUtils.java:82]: javadoc: it should end with period(.)

File: src/org/omegat/util/net/HttpConnectionUtils.java [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+    public static String getURL(URL url) throws IOException {
+        URLConnection urlConn = url.openConnection();
+        urlConn.setConnectTimeout(TIMEOUT_MS);
+        urlConn.setReadTimeout(TIMEOUT_MS);
+        try (InputStream in = urlConn.getInputStream()) {
+            return IOUtils.toString(in, StandardCharsets.UTF_8);
+        }
+    }
+
+    public static void downloadZipFileAndExtract(URL url, File dir, List<String> expectedFiles) throws IOException {
```
Comment (miurahr) [inline src/org/omegat/util/net/HttpConnectionUtils.java:93]: need javadoc comment

File: src/org/omegat/util/net/HttpConnectionUtils.java [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+                throw new FileNotFoundException("Could not extract expected files from zip; expected: "
+                        + expectedFiles + ", extracted: " + extracted);
+            }
+        } catch (IllegalArgumentException ex) {
+            throw new FlakyDownloadException(ex);
+        }
+    }
+
+    /**
+     * Downloads a binary file from a URL
```
Comment (miurahr) [inline src/org/omegat/util/net/HttpConnectionUtils.java:109]: javadoc: need period

File: src/org/omegat/util/net/HttpConnectionUtils.java [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+        }
+        try (InputStream in = conn.getInputStream()) {
+            return IOUtils.toByteArray(in);
+        } finally {
+            conn.disconnect();
+        }
+    }
+
+    /**
+     * Method call without additional headers for possible calls from plugins.
```
Comment (miurahr) [inline src/org/omegat/util/net/HttpConnectionUtils.java:168]: javadoc: need @param here

**Review by miurahr — COMMENTED** [review]
There are several place I can improve javadoc comments.

File: src/org/omegat/util/StaticUtils.java [hunk 5]
```diff
@@ -445,15 +443,14 @@ private static void replaceGlobs(StringBuilder haystack, String needle, String r
     }
 
     /**
-     * Download a file to memory
+     * Download a file to memory.
+     * @Deprecated
+     * This method is replaced to HttpConnectionUtils.getURL(url)
+     * No longer accept timeout.
```
Comment (amake) [inline src/org/omegat/util/StaticUtils.java:449]: Was there a reason to get rid of the timeout parameter?

File: src/org/omegat/util/net/HttpConnectionUtils.java [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+        }
+    }
+
+    /**
+     * Downloads a binary file from a URL
+     * @param jarFileURL HTTP URL of the file to be downloaded
+     * @param saveFilePath path of the file
+     * @throws IOException
+     */
+    public static boolean downloadBinaryFile(final URL jarFileURL, final Map<String, String> headers,
```
Comment (amake) [inline src/org/omegat/util/net/HttpConnectionUtils.java:114]: Where did this method come from? It isn't used anywhere.

Is this something you need for #69?

**Review by amake — COMMENTED** [review]
Sorry for the delay on this. I just had some minor questions.

Also I wonder if the package `org.omegat.util.net` isn't too much. What about just `org.omegat.util`? Or even `org.omegat.net`?

File: src/org/omegat/util/StaticUtils.java (hunk 5 above)
Comment (miurahr) [reply src/org/omegat/util/StaticUtils.java:449]: Functions which use the function set a same timeout value so I remove it. Now there is a compat function so I'd like to revert to support timeout.

File: src/org/omegat/util/net/HttpConnectionUtils.java (hunk 6 above)
Comment (miurahr) [reply src/org/omegat/util/net/HttpConnectionUtils.java:114]: > Is this something you need for #69?

Yes, I plan to use it to download Jar file from repository.

File: src/org/omegat/util/StaticUtils.java (hunk 5 above)
Comment (miurahr) [reply src/org/omegat/util/StaticUtils.java]: Now accept timeout.

**Comment by amake** [comment]
Sorry for the delay on this.

I will commit this after the release of 5.5.0 shortly. Can I get your opinion on this last issue please:

> Also I wonder if the package `org.omegat.util.net` isn't too much. What about just `org.omegat.util`? Or even `org.omegat.net`?

**Comment by miurahr** [comment]
> Sorry for the delay on this.
> 
> I will commit this after the release of 5.5.0 shortly. Can I get your opinion on this last issue please:
> 
> > Also I wonder if the package `org.omegat.util.net` isn't too much. What about just `org.omegat.util`? Or even `org.omegat.net`?

OK, I'd like to move it to `org.omegat.util`.

**Comment by amake** [comment]
Thank you and sorry for the delay. I've squashed and committed to `master`.

**Pull request closed without merging**